core/rust/qdbr/src/parquet_read/decode.rs

```diff
@@ -1282,29 +1263,171 @@ fn decode_page0<T: Pushable>(
     if def_levels.is_empty() {
         return Ok(None);
     }
 
     let iter = hybrid_rle::Decoder::new(def_levels, 1);
     let iter = HybridDecoderBitmapIter::new(iter, count);
-    let selected_rows = get_selected_rows(page);
-    let iter = FilteredHybridRleDecoderIter::new(iter, selected_rows);
     Ok(Some(iter))
 }
 
-pub fn get_selected_rows(page: &DataPage) -> VecDeque<Interval> {
-    page.selected_rows()
-        .unwrap_or(&[Interval::new(0, page.num_values())])
-        .iter()
-        .copied()
-        .collect()
+#[allow(clippy::while_let_on_iterator)]
+fn decode_array_page<T: DataPageSlicer>(
+    page: &DataPage,
+    row_lo: usize,
+    row_hi: usize,
+    slicer: &mut T,
+    buffers: &mut ColumnChunkBuffers,
+) -> ParquetResult<()> {
+    let count = slicer.count();
+    buffers.aux_vec.reserve(count * ARRAY_AUX_SIZE)?;
+    buffers.data_vec.reserve(slicer.data_size())?;
+
+    let (rep_levels, def_levels, _) = split_buffer(page)?;
+
+    let max_rep_level = page.descriptor.max_rep_level;
+    let max_def_level = page.descriptor.max_def_level;
+
+    let mut levels_iter = LevelsIterator::try_new(
+        page.num_values(),
+        max_rep_level,
+        max_def_level,
+        rep_levels,
+        def_levels,
+    )?;
+    let mut row = 0;
+    let mut skip_count = row_lo;
+    while let Some(levels) = levels_iter.next_levels() {
+        let levels = levels?;
+        // process levels accumulated for an array
+        if skip_count == 0 {
+            append_array(
+                &mut buffers.aux_vec,
+                &mut buffers.data_vec,
+                max_rep_level as u32,
+                max_def_level as u32,
+                &levels.rep_levels,
+                &levels.def_levels,
+                slicer,
+            )?;
+        } else {
+            skip_array(slicer, max_def_level as u32, &levels.def_levels);
+            skip_count -= 1;
+        }
+        row += 1;
+        if row >= row_hi {
+            return Ok(());
+        }
+    }
+    Ok(())
+}
+
+fn append_array<T: DataPageSlicer>(
+    aux_mem: &mut AcVec<u8>,
+    data_mem: &mut AcVec<u8>,
+    max_rep_level: u32,
+    max_def_level: u32,
+    rep_levels: &[u32],
+    def_levels: &[u32],
+    slicer: &mut T,
+) -> ParquetResult<()> {
+    if def_levels.len() == 1 && def_levels[0] == 0 {
+        append_array_null(aux_mem, data_mem)?;
+    } else {
+        let shape_size = align8b(4 * max_rep_level as usize);
+        let value_size = shape_size + 8 * rep_levels.len();
+
+        aux_mem.extend_from_slice(&data_mem.len().to_le_bytes())?;
+        aux_mem.extend_from_slice(&value_size.to_le_bytes())?;
+
+        // first, calculate and write shape
+        let mut shape = [0_u32; ARRAY_NDIMS_LIMIT];
+        calculate_array_shape(&mut shape, max_rep_level, rep_levels);
+        let mut num_elements: usize = 1;
+        for &dim in shape.iter().take(max_rep_level as usize) {
+            num_elements *= dim as usize;
```

Comment:
shall we add overflow check?

Reply:
There is the below `if num_elements != def_levels.len()` check. In case when an overflow occurs, the numbers won't match and the check will fail, so that we'll return an error to the user.